phantom_token.c

```diff
@@ -488,211 +495,102 @@ static ngx_int_t handler(ngx_http_request_t *request)
+        if (cache_data.data == NULL)
+        {
+            ngx_log_error(NGX_LOG_ERR, request->connection->log, 0, "Failed to allocate memory for cache data");
+        }
+        else
         {
             ngx_read_file(&request->cache->file, cache_data.data, cache_data.len, request->cache->body_start);
-
             jwt_start = cache_data.data;
+            jwt_len = request->headers_out.content_length_n;
         }
     }
     else
     {
-        jwt_start = request->header_end + sizeof("\r\n") - 1; // FIXME: Won't work if JWT is large
+        use_buffer_response = true;
     }
+#else
+    use_buffer_response = true;
+#endif
 
-    if (jwt_start == NULL)
+    // When caching is disabled the JWT is always read from upstream buffers
+    if (use_buffer_response)
     {
-        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request->connection->log, 0,
-                       "Failed to obtain JWT from introspection response or, if applicable, cache");
+        // Previous logic that could cause an end of buffer read:
+        // jwt_start = request->header_end + sizeof("\r\n") - 1;
+        // jwt_len = request->headers_out.content_length_n;
+
+        if (request->upstream->buffer.last == request->upstream->buffer.end)
+        {
+            ngx_log_error(NGX_LOG_ERR, request->connection->log, 0, "Buffer is maxed out, check the proxy_buffer_size configuration setting");
+            module_context->status = NGX_HTTP_INTERNAL_SERVER_ERROR;
+            return NGX_ERROR;
+        }
 
+        jwt_start = request->upstream->buffer.pos;
+        jwt_len = request->upstream->buffer.last - request->upstream->buffer.pos;
+    }
+
+    if (jwt_start == NULL || jwt_len == 0)
+    {
+        ngx_log_error(NGX_LOG_WARN, request->connection->log, 0, "Failed to obtain JWT from introspection response or cache");
         module_context->done = 1;
         module_context->status = NGX_HTTP_UNAUTHORIZED;
-
         return introspection_subrequest_status_code;
     }
-#else
-    jwt_start = request->header_end + sizeof("\r\n") - 1; // FIXME: Won't work if JWT is large
-#endif
-
-    size_t jwt_len = request->headers_out.content_length_n;
-    size_t bearer_jwt_len = BEARER_SIZE + jwt_len;
 
+    bearer_jwt_len = BEARER_SIZE + jwt_len;
     module_context->jwt.len = bearer_jwt_len;
     module_context->jwt.data = ngx_pnalloc(request->pool, bearer_jwt_len);
 
     if (module_context->jwt.data == NULL)
     {
+        ngx_log_error(NGX_LOG_ERR, request->connection->log, 0, "Failed to allocate memory for JWT token");
         module_context->done = 1;
         module_context->status = NGX_HTTP_UNAUTHORIZED;
-
         return introspection_subrequest_status_code;
     }
 
-    u_char *p = ngx_copy(module_context->jwt.data, BEARER, BEARER_SIZE);
-
+    p = ngx_copy(module_context->jwt.data, BEARER, BEARER_SIZE);
     ngx_memcpy(p, jwt_start, jwt_len);
-
-    if (cache_data.len > 0)
-    {
-        ngx_pfree(request->pool, cache_data.data);
```

Comment:
The free block is not needed and we do not free other memory allocations. Also I verified no leaks with valgrind after a few thousand API requests.